README.md
ustreasuries [![DOI](https://zenodo.org/badge/5738/grfiv/ustreasuries.svg)](https://zenodo.org/badge/latestdoi/5738/grfiv/ustreasuries)
=======================================================================================================================================

### US Treasury Bond Data

-   Real-Time downloads of
    -   daily Constant-Maturity Treasury (CMT) yields from 1962 to the most-recently completed business day
    -   daily prices for 400 Treasury securities (FedInvest) from 2010 to the most-recently completed business day

### Bond Analysis

Several models such as Nelson Seigel and Svensson are provided

**Veronesi** <sup>[Veronesi](#Veronesi)</sup> examples derived with code and live bond data

### Derivative Analysis

Drawing primarily on **Hull, 7th edition** <sup>[Hull](#Hull)</sup> *ustreasuries* includes many options-pricing models, all the Greeks and a number of utility functions, all of which have examples from Hull which demonstrate that they work correctly.

------------------------------------------------------------------------

### Wiki

See the [GitHub Wiki](https://github.com/grfiv/ustreasuries/wiki) for examples of the use of all the functions.

### Functions

-   **Treasury Rate Data**
    -   **CMTrates** downloads real-time daily CMT data from 1962
        -   **PrintYieldCurves** prints one or more CMT yield curves
        -   **APY** converts Constant-Maturity Treasury (CMT) yields to Annualized Percentage Yields (APY)
    -   **FedInvestData** downloads real-time treasury bond price data from 2010
        -   **CoerceFedInvest\_xts** Turn FedInvest data into a time series
-   **Black-Scholes-Merton**
    -   **EuroCall** Calculate the price of a European call option with or without dividends
    -   **EuroPut** Calculate the price of a European put option with or without dividends
    -   **EuroCallVol** Implied Volatility for a European Call option
    -   **EuroPutlVol** Implied Volatility for a European Put option
-   **Greeks**
    -   **DeltaCall** Amount call-option price changes given a change in asset price
    -   **DeltaPut** Amount put-option price changes given a change in asset price
    -   **ThetaCall** the decay in the value of a call or a portfolio of calls as time passes
    -   **ThetaPut** the decay in the value of a put or a portfolio of puts as time passes
    -   **Gamma** the change in Delta with respect to asset price
    -   **Vega** the sensitivity to changes in the volatility of the underlying
    -   **RhoCall** the sensitivity to changes in the risk-free rate of return
    -   **RhoPut** the sensitivity to changes in the risk-free rate of return
-   **Utility Functions**
    -   **Bonds**
        -   **NSzeros** Convert Nelson Seigel and Svensson rates to Z(0, T)
        -   **Zbootstrap** Derive Z(0, T\_i) using the bootstrap method
        -   **spot\_rate** Derive spot rate from Z(0, T) and T
    -   **CAGR**
        -   **CAGR** Calculate Compound Annual Growth Rate; geometric or continuous
        -   **r\_continuous** Convert from discrete to continuous
        -   **r\_discrete** Convert from continuous to discrete
    -   **Discount Factors**
        -   **discount\_factor** Calculate discount factor Z(t, T)
        -   **interest\_rate** Calculate annualized interest rate r(t, T) from a discount factor Z(t, T)
    -   **Put/Call Parity**
        -   **CallParity** Convert from a put-option price using put/call parity
        -   **PutParity** Convert from a call-option price using put/call parity
    -   **Risk Neutral/Forwards**
        -   **RiskNeutralProb** Binomial tree risk-neutral probability
        -   **ForwardPrice** Forward price with or without income or yield
        -   **ForwardRate** Forward rate from Time1 to Time2 (discrete compounding)
    -   **Options**
        -   **IntrinsicValueCall** / **IntrinsicValuePut** the in-the-money portion of an option's premium
        -   **TimeValueCall** / **TimeValuePut** Price = Intrinsic + Time
        -   **InTheMoneyCall** / **InTheMoneyPut** Is an option in the money?
    -   **Equities**
        -   **SP500** Daily S&P 500 data from 1950
        -   **SP500TR** Daily S&P 500 Total Return data from 1988
-   **Installed but not yet tested or documented**
    -   Digital
        -   **CashCall**
        -   **CashPut**
        -   **AssetCall**
        -   **AssetPut**
    -   Greeks
        -   **RhoFuturesCall**
        -   **RhoFuturesPut**
        -   **RhoFXCall**
        -   **RhoFXPut**
    -   American
        -   **American\_Put\_Binomial**
        -   **American\_Call\_Dividend**

#### All of the derivatives functions have examples drawn from **Hull, 7th edition** <sup>[Hull](#Hull)</sup> to demonstrate their correctness.

See <https://github.com/grfiv/BlackScholesMerton> for these functions written in Python and Excel VBA

See also <http://www.philadelphia-reflections.com/topic/230.htm>

Installation
------------

We're not on CRAN yet; get the development version from GitHub:

``` r
# see https://github.com/hadley/devtools for the
# best procedure to install *devtools* on your
# system; Windows in particular has somewhat
# complicated requirements

devtools::install_github("grfiv/treasuries")

# Notes: 
#    1. Add 'build_vignettes=TRUE' to include vignettes 
#       (recommended, but a current version of pandoc is required)
#    2. add 'auth_token="..."' if you get a 404
#       contact the author for this
#    3. if you receive a message about corrupt databases or fetch(key), 
#       restarting R will fix the problem;
#       these appear to be issues with devtools 1.10.0.9000
```

References
----------

<a name="Back">\[Back\]</a>: Back, K., *A Course in Derivative Securities* 2005 Springer Finance ISBN 9783540253734

<a name="Consiglio">\[Consiglio\]</a>: Consiglio A. and Guirreri S.S. *Simulating the Term Structure of Interest Rates with arbitrary marginals*. International Journal of Risk Assessment and Management, 15(4), September 2011.

<a name="Hull">\[Hull\]</a>: Hull, J., *Options, Futures and Other Derivatives Seventh Edition* 2008 Pearson/Prentice Hall ISBN 9780136015864

<a name="CMT">\[CMT\]</a>: US Dept. of Treasury Daily Treasury Yield Curve Rates
<https://www.treasury.gov/resource-center/data-chart-center/interest-rates/Pages/TextView.aspx?data=yield>

<a name="H15">\[H15\]</a>: US Federal Reserve Board Data Download Program
<http://www.federalreserve.gov/datadownload/Choose.aspx?rel=H15>

<a name="Varma">\[Varma\]</a>:
Varma, J. *jrvFinance* v1.03 (<https://github.com/jrvarma/jrvFinance>)

<a name="Veronesi">\[Veronesi\]</a>: Veronesi P., *Fixed Income Securities* 2010 John Wiley & Sons ISBN 9780470109106

<a name="MDC">\[MDC\]</a>: Wall Street Journal Market Data Center
<http://online.wsj.com/mdc/public/page/mdc_bonds.html?mod=mdc_topnav_2_3010>


NEWS.md
# ustreasuries 0.0.0.9006

-------------------------------------------------------------------------------
--------

## Seventh Development Release 0.0.0.9006   
* ### Added vignette *veronesi-ch02*, all the examples and figures from Chapter 2 of Veronesi   
* ### Added vignette *veronesi-ch02-Nelson-Seigel*, the appendix to Veronesi Chapter 2 showing the Nelson Seigel and Svensson term structure models      
* ### Added *Zbootstrap* function
* ### Modified *FedInvestData* function to start at the date following the last date in the historical file
* ### Added *data-raw/FedInvestHistory_bring_to_current.R* to "top off" the FedInvest historical dataset so that as few POSTS as possible have to be made to bring it current
* ### Added *SP500* and *SP500TR* to download S&P 500 data
* ### Added *spot_rate* function
* ### Added *CoerceFedInvest_xts* function


-------------------------------------------------------------------------------
--------

## Sixth Development Release 0.0.0.9005
* ### Extensive miscellaneous cleanup and refactoring
* ### Added rudimentary bond math

-------------------------------------------------------------------------------
--------

## Fifth Development Release 0.0.0.9004
* ### Added References in README and project documentation
* ### Changed **OptionGamma** to **Gamma**
* ### Changed **USTreasuryRates** to **CMTrates**
* ### Changed **CAGRd** to **CAGR**, added geometric & continuous
* ### Added discount factor functions
    * ### **discount_factor**, **interest_rate**
* ### Cited my use of Prof. Jayanth R Varma's *jrvFinance* package Lic: GPL (>=2)
* ### **CMTrates** modified to to add attr(..., "data.source") <- "CMT"
* ### Added **xts (>= 0.9.7)** to DESCRIPTION Imports
* ### Added **Nelson.Siegel**, **NSrates**, **NSzeros**, **Nelson.Siegel.plot**, **NSzeros.plot**
* ### Added **FedInvestData** Return treasury bond price data from 2010

-------------------------------------------------------------------------------
--------

## Fourth Development Release 0.0.0.9003
* ### Added Implied Volatlity **EuroCallVol** / **EuroPutVol**
* ### Added Time Value  **TimeValueCall** / **TimeValuePut**
* ### Updated black-scholes-merton vignette
* ### Updated GitHub Wiki

-------------------------------------------------------------------------------
--------

## Third Development Release 0.0.0.9002
* ### Added Greeks
    * **DeltaCall** / **DeltaPut**
    * **ThetaCall** / **ThetaPut**
    * **OptionGamma**
    * **Vega** 
    * **RhoCall** / **RhoPut**
* ### Added Utilities
    * **IntrinsicValueCall** / **IntrinsicValuePut**
    * **InTheMoneyCall** / **InTheMoneyPut**
* ### Added examples to ustreasuries.R
* ### Updated black-scholes-merton vignette
* ### Updated GitHub Wiki

-------------------------------------------------------------------------------
--------

## Second Development Release 0.0.0.9001

* ### Added all the Black Scholes Merton functions
    * **Black-Scholes-Merton**
        * **EuroCall** Calculate the price of a European call option with or without dividends
        * **EuroPut** Calculate the price of a European put option with or without dividends
    * **Utility Functions**
        * **CAGR**
            * **CAGRd** Calculate discrete Compound Annual Growth Rate
            * **r_continuous** Convert from discrete to continuous CAGR
            * **r_discrete** Convert from continuous to discrete CAGR
        * **Put/Call Parity**
            * **CallParity** Convert from a put-option price using put/call parity
            * **PutParity** Convert from a call-option price using put/call parity
        * **Risk Neutral/Forwards**
            * **RiskNeutralProb** Binomial tree risk-neutral probability
            * **ForwardPrice** Forward price with or without income or yield
            * **ForwardRate** Forward rate from Time1 to Time2 (discrete compounding)
    * **Installed but not yet undocumented**
        * **CashCall**
        * **CashPut**
        * **AssetCall**
        * **AssetPut**
        * **American_Put_Binomial**
        * **DeltaCall**
        * **DeltaPut**
        * **OptionGamma**
        * **ThetaCall**
        * **ThetaPut**
        * **Vega**
        * **RhoFuturesCall**
        * **RhoFuturesPut**
        * **RhoFXCall**
        * **RhoFXPut**
        * **RhoCall**
        * **RhoPut**
        * **EuroCallVol**
        * **EuroPutlVol**
        * **Black_Scholes_Call_Implied_Vo**l
        * **Black_Scholes_Put_Implied_Vol**
        * **ImpVolCall**
        * **ImpVolPut**
        * **American_Call_Dividend**

* ### Modified *yield-curves* vignette to show last three years
* ### Added two vignettes for the derivatives functions
    * *utilities* examples of the utility functions
    * *black-scholes-merton* examples of options pricing
    
-------------------------------------------------------------------------------   
------------------------   

## Initial Development Release 0.0.0.9000

* ### Three fully-documented functions (roxygen2)
    * **USTreasuryRates** downloads a data.frame with daily data from 1962
    * **PrintYieldCurves** prints one or more yield curves
    * **APY** converts Constant-Maturity Treasury (CMT) yields to Annualized Percentage Yields (APY)
    
* ### Three vignettes
    * **cmt-rates** a description of Constant Maturity and Annualized Percentage rates
    * **yield-curves** examples of downloading the data and printing yield curves for interesting periods in recent financial history
    * **plot-10year** a plot of the 10-year from 1962 to present
    
* ### Unit Tests (testthat) *devtools::test()*
    * **USTreasuryRates** Test data download : ...............
    * **APY**             Test APY : ..
    
* ### R CMD RStudio Build>Check Package, *devtools::check()*, *devtools::build_win()* 
    * No errors, warnings or notes
    * "Status: OK, R CMD check succeeded""
    
* ### Initial upload to GitHub
    * *devtools::install_github("grfiv/ustreasuries", build_vignettes=TRUE)*    

R/ustreasuries.R
#' US Treasury bond analysis
#'
#' The \emph{ustreasuries} package downloads daily Constant-Maturity Treasury
#' (CMT) yields and provides visualizations & analytics that use that data
#' including all the 'greeks' for derivative analysis.
#'
#' @section GitHub Wiki:
#' \url{https://github.com/grfiv/ustreasuries/wiki}
#'
#' @section Vignettes:
#' \itemize{
#'     \item \bold{summary} Overview of package functionality
#'     \item \bold{veronesi-ch02} Examples and figures from Veronesi, Ch2
#'     \item \bold{cmt-rates} a description of Constant Maturity and
#'     Annualized Percentage rates
#'     \item \bold{yield-curves} examples of downloading the data and printing
#'     yield curves for interesting periods in recent financial history.
#'     \item \bold{plot-10year} a plot of the 10-year from 1962 to present
#'     \item \bold{utilities} examples of the utility functions
#'     \item \bold{black-scholes-merton} examples of options pricing
#' }
#'
#' @section Functions:
#' \itemize{
#'
#' \item \bold{Treasury Rate Data}
#'     \itemize{
#'        \item \bold{CMTrates} downloads real-time daily CMT data from 1962
#'        \itemize{
#'            \item \bold{PrintYieldCurves} prints one or more yield curves
#'            \item \bold{APY} converts Constant-Maturity Treasury (CMT) yields to
#'     Annualized Percentage Yields (APY)
#'     }
#'     \item \bold{FedInvestData} downloads real-time treasury bond price data from 2010
#'     \itemize{
#'          \item \bold{CoerceFedInvest_xts} Turn FedInvest data into a time series
#'     }
#'    }
#'
#'  \item \bold{Black Scholes Merton}
#'      \itemize{
#'          \item \bold{EuroCall} Calculate the price of a European call option with or without dividends
#'          \item \bold{EuroPut} Calculate the price of a European put option with or without dividends
#'          \item \bold{EuroCallVol} Implied Volatility for a European call option
#'          \item \bold{EuroPutlVol} Implied Volatility for a European put option
#'      }
#'
#'  \item \bold{Greeks}
#'      \itemize{
#'          \item \bold{DeltaCall} Amount call-option price changes for a change in asset price
#'          \item \bold{DeltaPut} Amount put-option price changes for a change in asset price
#'          \item \bold{ThetaCall} the decay in the value of a call or a portfolio of calls as time passes
#'          \item \bold{ThetaPut} the decay in the value of a put or a portfolio of puts as time passes
#'          \item \bold{Gamma} the change in Delta with respect to asset price
#'          \item \bold{Vega} the sensitivity to changes in the volatility of the underlying
#'          \item \bold{RhoCall} the sensitivity to changes in the risk-free rate of return
#'          \item \bold{RhoPut} the sensitivity to changes in the risk-free rate of return
#'      }
#'
#'  \item \bold{Utility Functions}
#'  \itemize{
#'          \item \bold{Bonds}
#'              \itemize{
#'                  \item \bold{NSzeros} Convert Nelson Seigel and Svensson rates to Z(0, T)
#'                  \item \bold{Zbootstrap} Derive Z(0, T_i) using the bootstrap method
#'                  \item \bold{spot_rate} Derive spot rate from Z(0, T) and T
#'              }
#'      }
#'      \itemize{
#'          \item \bold{CAGR}
#'              \itemize{
#'                  \item \bold{CAGR} Calculate Compound Annual Growth Rate; geometric or continuous
#'                  \item \bold{r_continuous} Convert from discrete to continuous
#'                  \item \bold{r_discrete} Convert from continuous to discrete
#'              }
#'           \item \bold{Discount Factors}
#'              \itemize{
#'                  \item \bold{discount_factor} Calculate discount factor Z(t, T)
#'                  \item \bold{interest_rate} Calculate annualized interest rate r(t, T) from a discount factor Z(t, T)
#'              }
#'          \item \bold{Put/Call Parity}
#'          \itemize{
#'              \item \bold{CallParity} Convert from a put-option price using put/call parity
#'              \item \bold{PutParity} Convert from a call-option price using put/call parity
#'              }
#'          \item \bold{Risk Neutral/Forwards}
#'              \itemize{
#'                  \item \bold{RiskNeutralProb} Binomial tree risk-neutral probability
#'                  \item \bold{ForwardPrice} Forward price with or without income or yield
#'                  \item \bold{ForwardRate} Forward rate from Time1 to Time2 (discrete compounding)
#'              }
#'          \item \bold{Options}
#'              \itemize{
#'                  \item \bold{IntrinsicValueCall} The in-the-money portion of a call option's premium
#'                  \item \bold{IntrinsicValuePut} The in-the-money portion of a put option's premium
#'                  \item \bold{TimeValueCall} Price = Intrinsic + Time
#'                  \item \bold{TimeValuePut} Price = Intrinsic + Time
#'                  \item \bold{InTheMoneyCall} Is a call in the money?
#'                  \item \bold{InTheMoneyPut} Is a put in the money?
#'              }
#'          \item \bold{Equities}
#'              \itemize{
#'                  \item \bold{SP500} Daily S&P 500 data from 1950
#'                  \item \bold{SP500TR} Daily S&P 500 Total Return data from 1988
#'                  }
#'      }
#'
#'  \item \bold{Installed but not yet tested and documented}
#'      \itemize{
#'          \item Digital
#'          \itemize{
#'              \item \bold{CashCall}
#'              \item \bold{CashPut}
#'              \item \bold{AssetCall}
#'              \item \bold{AssetPut}
#'            }
#'          \item Greeks
#'          \itemize{
#'              \item \bold{RhoFuturesCall}
#'              \item \bold{RhoFuturesPut}
#'              \item \bold{RhoFXCall}
#'              \item \bold{RhoFXPut}
#'          }
#'          \item American
#'          \itemize{
#'              \item \bold{American_Put_Binomial}
#'              \item \bold{American_Call_Dividend}
#'          }
#'      }
#' }
#'
#' @section Installation:
#'     \code{## Windows users STOP: see https://github.com/hadley/devtools}
#'
#'     \code{install.packages("devtools") }
#'
#'     \code{devtools::install_github("grfiv/ustreasuries",
#'                                    build_vignettes=TRUE)}
#'
#' \code{## NOTE: If you get a 404 on installation, contact the author for the 'auth_token' parameter}
#'
#'     (restart R)
#'
#' @examples
#' # example of US Treasury rate data download
#' # =========================================
#' all_data <- CMTrates()
#' tail(all_data)
#' # ==================================================
#' # last row displayed should be for last business day
#' # ==================================================
#'
#' # example of option pricing
#' # =========================
#' # Hull 7th edition Ch 13 P 294
#' Stock     <- 42    # S_0
#' Exercise  <- 40    # K
#' Time      <- 0.50  # T
#' Interest  <- 0.10  # r
#' Yield     <- 0     # q
#' sigma     <- 0.20
#' ans <- EuroCall(Stock, Exercise, Time, Interest, Yield, sigma)
#' round(ans,2) # 4.76
#'
#' @references
#' Back, K., \emph{A Course in Derivative Securities} 2005 Springer Finance ISBN 9783540253734
#'
#' Consiglio A. and Guirreri S.S. \emph{Simulating the Term Structure of Interest Rates with
#' arbitrary marginals}. International Journal of Risk Assessment and Management, 15(4),
#' September 2011.
#'
#' Hull, J., \emph{Options, Futures and Other Derivatives Seventh Edition} 2008 Pearson/Prentice Hall ISBN 9780136015864
#'
#' US Dept. of Treasury Daily Treasury Yield Curve Rates
#' \url{https://www.treasury.gov/resource-center/data-chart-center/interest-rates/Pages/TextView.aspx?data=yield}
#'
#' US Federal Reserve Board Data Download Program
#' \url{http://www.federalreserve.gov/datadownload/Choose.aspx?rel=H15}
#'
#' Varma, J. \emph{jrvFinance} v1.03
#' \url{https://github.com/jrvarma/jrvFinance}
#'
#' Veronesi P., \emph{Fixed Income Securities} 2010 John Wiley & Sons ISBN 9780470109106
#'
#' Wall Street Journal Market Data Center
#' \url{http://online.wsj.com/mdc/public/page/mdc_bonds.html?mod=mdc_topnav_2_3010}
#'
#' @author
#' George Fisher \email{GeorgeRFisher@gmail.com}
#'
#' @docType package
#' @name ustreasuries
NULL
